core/src/sumeragi/mod.rs

```diff
@@ -1574,20 +1565,22 @@ pub mod message {
     #[non_exhaustive]
     pub struct TransactionForwarded {
         /// Transaction that is forwarded from a client by a peer to the leader
-        pub transaction: VersionedAcceptedTransaction,
+        pub transaction: VersionedTransaction,
         /// `PeerId` of the peer that forwarded this transaction to a leader.
         pub peer: PeerId,
     }
 
     impl TransactionForwarded {
         /// Constructs `TransactionForwarded` message.
         pub fn new(
-            transaction: &VersionedAcceptedTransaction,
-            peer: &PeerId,
+            transaction: VersionedAcceptedTransaction,
+            peer: PeerId,
         ) -> TransactionForwarded {
             TransactionForwarded {
-                transaction: transaction.clone(),
-                peer: peer.clone(),
+                // Converting into non-accepted transaction because it's not possible
+                // to guarantee that the sending peer checked transaction limits
+                transaction: transaction.into(),
```

Comment:
I think we should somehow forbid sending anything other than `Transaction`/`VersionedTransaction`

Reply:
yes, I added this in validation step as well as downgraded transactions in sumeragi messages to be plain `VersionedTransaction`.

Maybe we could make some trait scheme to prevent these situations to compile? Like require all structs in messages to implement some trait like `MessagePayload`. This would, I think, require some macro but should be more robust

Reply:
Yes, robustness should be taken even if it becomes a little complicated.
Could you explain what `MessagePayload` method will ensure the sanity in the message handling?

Reply:
That should be left to another PR though

Reply:
I agree with Sato.  This should be discussed separately, because a `trait` isn't the only solution to this problem.

Reply:
of course, I didn't mean to deal with this here